mysql/resource_default_roles.go

```diff
@@ -0,0 +1,211 @@
+
+	if err := alterUserDefaultRoles(ctx, db, user, host, roles); err != nil {
+		return diag.Errorf("failed to create user default roles: %v", err)
+	}
+
+	d.SetId(fmt.Sprintf("%s@%s", user, host))
+
+	return nil
+}
+
+func UpdateDefaultRoles(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	db, err := getDatabaseFromMeta(ctx, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if err := checkDefaultRolesSupport(ctx, meta); err != nil {
+		return diag.Errorf("cannot use default roles: %v", err)
+	}
+
+	if d.HasChange("roles") {
+		user := d.Get("user").(string)
+		host := d.Get("host").(string)
+		roles := getRolesFromData(d)
+
+		if err := alterUserDefaultRoles(ctx, db, user, host, roles); err != nil {
+			return diag.Errorf("failed to update user default roles: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func ReadDefaultRoles(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	db, err := getDatabaseFromMeta(ctx, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if err := checkDefaultRolesSupport(ctx, meta); err != nil {
+		return diag.Errorf("cannot use default roles: %v", err)
+	}
+
+	stmtSQL := "SELECT default_role_user FROM mysql.default_roles WHERE user = ? AND host = ?"
+
+	log.Println("Executing statement:", stmtSQL)
+
+	rows, err := db.QueryContext(ctx, stmtSQL, d.Get("user").(string), d.Get("host").(string))
+	if err != nil {
+		return diag.Errorf("failed to read user default roles from DB: %v", err)
+	}
+	defer rows.Close()
+
+	var defaultRoles = make([]string, 0)
+	for rows.Next() {
+		var role string
+		err := rows.Scan(&role)
+		if err != nil {
+			return diag.Errorf("failed scanning default roles: %v", err)
+		}
+		defaultRoles = append(defaultRoles, role)
+	}
+
+	if rows.Err() != nil {
+		return diag.Errorf("failed getting rows: %v", rows.Err())
+	}
+
+	d.Set("roles", defaultRoles)
+
+	return nil
+}
+
+func DeleteDefaultRoles(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	db, err := getDatabaseFromMeta(ctx, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if err := checkDefaultRolesSupport(ctx, meta); err != nil {
+		return diag.Errorf("cannot use default roles: %v", err)
+	}
+
+	user := d.Get("user").(string)
+	host := d.Get("host").(string)
+
+	if err := alterUserDefaultRoles(ctx, db, user, host, []string{}); err != nil {
+		return diag.Errorf("failed to remove user default roles: %v", err)
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func ImportDefaultRoles(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	userHost := strings.SplitN(d.Id(), "@", 2)
+
+	if len(userHost) != 2 {
+		return nil, fmt.Errorf("wrong ID format %s (expected USER@HOST)", d.Id())
+	}
+
+	d.Set("user", userHost[0])
+	d.Set("host", userHost[1])
+
+	var err error
+
+	readErr := ReadDefaultRoles(ctx, d, meta)
+	if readErr.HasError() {
+		err = fmt.Errorf("failed reading default roles: %v", err)
```

Comment:
The `err` parameter should probably be taken from readErr - shouldn't it?

Reply:
I changed the code here to iterate all Terraform diagnostics and return the first error. This I think is good enough since the `ReadDefaultRoles` function never returns more than one error. The `diag.Diagnostics` does not have a better print function afaik and printing it directly with `%v` was ugly.

Reply:
Yes, this is good enough. I only didn't want to have always "nil" there.